libraries/AP_LQR_Control/AP_LQR_Control.cpp

```diff
@@ -0,0 +1,409 @@
+    float u =  - (((_xtrack_fac*0.01)*q1*_crosstrack_error)+(sqrtf((float)((_q2_val*0.01)+(2*q1)))*(_vel_fac*0.01)*v_d));
+    
+    //Limit calculated lateral acceleration to maximum values
+    if (fabs(u) > (_max_latacc*0.01))
+    {
+        if(u > 0)
+            u = _max_latacc*0.01;
+        if(u < 0)
+            u = - (_max_latacc*0.01);
+    }
+    _latAccDem= u;
+
+    
+    // Waypoint capture status is always false during waypoint following
+    _WPcircle = false;
+
+    
+
+    _data_is_stale = false; // status are correctly updated with current waypoint data
+}
+
+// update L1 control for loitering
+void AP_LQR_Control::update_loiter(const struct Location &center_WP, float radius, int8_t loiter_direction)
+{
+    struct Location _current_loc;
+    Vector2f _groundspeed_vector = _ahrs.groundspeed_vector();
+
+    // scale loiter radius with square of EAS2TAS to allow us to stay
+    // stable at high altitude
+    radius = loiter_radius(radius);
+    float groundSpeed=_groundspeed_vector.length();
+    Vector2f location_difference=location_diff(center_WP, _current_loc);
+    
+    //Calculate rate change of heading of path
+    float si_p_dot=groundSpeed/location_difference.length();
+    _crosstrack_error = location_difference.length() - radius;
+    
+    //Set minimum turn radius equal to twice the groundspeed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    float u =0;
+    
+    // check if vehicle is not very far from the desired circular path
+    if (_crosstrack_error < (radius))
+    {
+        //Compute desired heading perpendicular
+        float si_p = ((_target_bearing_cd + 9000)*0.01);
+        //Compute adaptive gains
+        float q1= sqrtf((float)((_max_xtrack*0.01)/(fabs((_max_xtrack*0.01)-_crosstrack_error))));
+        //Compute velocity of approach towards desired path
+        float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
+        float temp_sin=sinf(radians(si - si_p));
+        float v_d= groundSpeed * temp_sin;
+        //Compute desired lateral acceleration
+        u = - (((_xtrack_fac*0.01)*q1*0.5*_crosstrack_error)+(sqrtf((float)((_q2_val*0.01)+(2*q1)))*(_vel_fac*0.01)*v_d)+groundSpeed*si_p_dot);
+    }
+    
+    //lead towards center if vehicle is very far from circular path until crosstrack error < radius
+    else
+    {
+        float si_p = (_target_bearing_cd*0.01);
+        float q1= sqrtf((float)((_max_xtrack*0.01)/(fabs((_max_xtrack*0.01)-_crosstrack_error))));
+        float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
+        float temp_sin=sinf(radians(si - si_p));
+        float v_d= groundSpeed * temp_sin;
+        u = - (((_xtrack_fac*0.01)*q1*0*_crosstrack_error)+(sqrtf((float)((_q2_val*0.01)+(2*q1)))*(_vel_fac*0.01)*v_d));
```

Comment:
Why the first term gets zeroed in this case?

Reply:
If cross track error is too large, we want to approach towards the center point to reduce the cross track error at first. Thus cross track value is not deciding the lateral acceleration. Our goal is to head towards the center.
Also large cross track error will force the vehicle to loiter on the spot.